### ISSUE TITLE
InferenceSession(GraphOptimizationLevel.ORT_ENABLE_BASIC) inserts Cast nodes with hard-coded name cause name conflict

### ISSUE THREAD

**Issue by JianhuiD** [microsoft/onnxruntime#4745]
**Describe the bug**
ONNXRT  InferenceSession(GraphOptimizationLevel.ORT_ENABLE_BASIC) inserted Cast nodes with hard-coded name(both node name and output ) cause name conflict

**Urgency**
High

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version:  1.4.0
- Python version:  3.6.8
- Visual Studio version (if applicable): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: M/A

**To Reproduce**
- Run InferenceSession(GraphOptimizationLevel.ORT_ENABLE_BASIC, as below)  on the attached model twice  - the 2nd runs on the generated-model  from 1st run.  2nd run emitted the error "onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : This is an invalid model. Error: Duplicate definition of name (CastDef_0)" 
    ```
    so = onnxrt.SessionOptions()
    so.graph_optimization_level = onnxrt.GraphOptimizationLevel.ORT_ENABLE_BASIC
    so.optimized_model_filepath = model_cf_filename
    onnxrt.InferenceSession(inputmodel, sess_options=so)
   ```
   The attached model contains Float16 Data, The root of this issue is that InferenceSession try to  insert Cast nodes, however, the inserted node's name(output) is hard-coded which cause name conflict.  And OnnxRT NOT not assume there is no "CastDef_X" output used in the original model.

- Attach the ONNX model to the issue (where applicable) to expedite investigation.
[test_model.zip](https://github.com/microsoft/onnxruntime/files/5052702/test_model.zip)

**Expected behavior**
Shall not see this failure

**Comment by jcwchen** [comment]
Hi @JianhuiD, 
Thank you for providing the details. I can reproduce this error by your model and code.
Let me dig deeper into this to see where the problem is.

**Comment by hariharans29** [comment]
Closed via #4811
